src/harness/testcases/grant_testcase.py

```diff
@@ -624,3 +624,245 @@ def test_WINNF_FT_S_GRA_28(self):
+      response = self._sas.Heartbeat(request)['heartbeatResponse'][0]
+      # Check the heartbeat response
+      self.assertEqual(response['cbsdId'], cbsd_id)
+      self.assertEqual(response['grantId'], grant_id)
+      self.assertFalse('transmitExpireTime' in response)
+      self.assertEqual(response['response']['responseCode'], 501)
+
+  @winnforum_testcase 
+  def test_WINFF_FT_S_GRA_4(self):
+    """Successful CBSD grant request.
+        No incumbent present in the PAL frequency range requested by the CBSD.
+    """
+
+    # Register device inside PPA
+    device_a = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_a.json')))
+    self._sas_admin.InjectFccId({'fccId': device_a['fccId']})
+    request = {'registrationRequest': [device_a]}
+    response = self._sas.Registration(request)['registrationResponse'][0]
+    # Check registration response
+    self.assertEqual(response['response']['responseCode'], 0)
+    cbsd_id = response['cbsdId']
+    del request, response
+    # Inject PAL Database Recoord
+    pal_database_record = json.load(
+        open(os.path.join('testcases', 'testdata', 'pal_database_record_0.json')))
+    pal_database_record['userId'] = device_a['userId']
+    self._sas_admin.InjectPalDatabaseRecord(pal_database_record)
+    # Inject PPA Zone
+    ppa_zone = json.load(
+        open(os.path.join('testcases', 'testdata', 'ppa_zone_0.json')))
+    ppa_id = self._sas_admin.InjectZoneData(ppa_zone)
+    # Inject cbsd_ids of ppa
+    cluster_list = {'ppaId' : ppa_id['zoneId'], 'cbsdIds': [cbsd_id]}
+    self._sas_admin.InjectClusterList(cluster_list)
+    # grant request for cbds 
+    grant_0 = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_0['cbsdId'] = cbsd_id
+    request = {'grantRequest': [grant_0]}
+    response = self._sas.Grant(request)['grantResponse'][0]
+    # Check grant response
+    self.assertEqual(response['cbsdId'], cbsd_id)
+    self.assertTrue(response['grantId'])
+    self.assertEqual(response['channelType'], 'PAL')
+    self.assertEqual(response['response']['responseCode'], 0)
+
+  @winnforum_testcase 
+  def test_WINFF_FT_S_GRA_5(self):
+    """Successful CBSD grant request.
+        Incumbent present in the PAL frequency range requested by the CBSD 
+        who is outside the protection zone.
+    """
+
+    # Register device outside protection zone
+    device_outside_esc_zone = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_a.json')))
+    self._sas_admin.InjectFccId({'fccId': device_outside_esc_zone['fccId']})
+    request = {'registrationRequest': [device_outside_esc_zone]}
+    response = self._sas.Registration(request)['registrationResponse'][0]
+    # Check registration response
+    self.assertEqual(response['response']['responseCode'], 0)
+    cbsd_id = response['cbsdId']
+    del request, response
+    # Inject PAL Database Recoord
+    pal_database_record = json.load(
```

Comment:
thank you, it is a good point but then it is better to update Admin API document in the section of PAL database record to be two steps:
- inject census tract zone(by inject zone data) and to get the returned zone Id as census tract Id to build the PalId and the license area extent of the PAL database record.
- inject the PAL database record.
right?